packages/agent/src/storage/deferredCredentialStore.ts

```diff
@@ -0,0 +1,120 @@
+import { type Agent, type Kms, utils } from '@credo-ts/core'
+import type {
+  OpenId4VciDeferredCredentialResponse,
+  OpenId4VciMetadata,
+  OpenId4VciRequestTokenResponse,
+} from '@credo-ts/openid4vc'
+import { useMemo } from 'react'
+import { getWalletJsonStore, useWalletJsonRecord } from '../storage'
+
+export interface DeferredCredential {
+  id: string
+  createdAt: string
+  lastCheckedAt: string
+  lastErroredAt?: string
+  clientId?: string
+  response: OpenId4VciDeferredCredentialResponse
+  issuerMetadata: OpenId4VciMetadata
+  accessToken: Omit<OpenId4VciRequestTokenResponse, 'dpop'> & {
+    dpop?: Omit<NonNullable<OpenId4VciRequestTokenResponse['dpop']>, 'jwk'> & {
+      jwk: Kms.PublicJwk['jwk']
+    }
+  }
+}
+
+interface DeferredCredentialRecord {
+  deferredCredentials: DeferredCredential[]
+}
+
+const _deferredCredentialStorage = getWalletJsonStore<DeferredCredentialRecord>(
+  'EASYPID_OID4VCI_DEFERRED_CREDENTIAL_RECORD'
+)
+export const deferredCredentialStorage = {
+  recordId: _deferredCredentialStorage.recordId,
+  addDeferredCredential: async (agent: Agent, deferredCredential: DeferredCredential) => {
+    const record = await _deferredCredentialStorage.get(agent)
+    if (!record) {
+      await _deferredCredentialStorage.store(agent, {
+        deferredCredentials: [deferredCredential],
+      })
+    } else {
+      record.deferredCredentials.push(deferredCredential)
+      await _deferredCredentialStorage.update(agent, record)
+    }
+    return deferredCredential
+  },
+  updateDeferredCredential: async (
+    agent: Agent,
+    updatedDeferredCredential: Partial<DeferredCredential> & Pick<DeferredCredential, 'id'>
+  ) => {
+    const record = await _deferredCredentialStorage.get(agent)
+    if (!record) {
+      throw new Error('No deferred credential record found')
+    }
+
+    const deferredCredential = record.deferredCredentials.findIndex((d) => d.id === updatedDeferredCredential.id)
+    if (deferredCredential === -1) {
+      throw new Error(`Deferred credential with id ${updatedDeferredCredential.id} not found`)
+    }
+
+    record.deferredCredentials[deferredCredential] = {
+      ...record.deferredCredentials[deferredCredential],
+      ...updatedDeferredCredential,
+    }
+
+    await _deferredCredentialStorage.update(agent, record)
+    return updatedDeferredCredential
+  },
+  deleteDeferredCredential: async (agent: Agent, id: string) => {
+    const record = await _deferredCredentialStorage.get(agent)
+    if (!record) {
+      throw new Error('No deferred credential record found')
+    }
+
+    record.deferredCredentials = record.deferredCredentials.filter((d) => d.id !== id)
+    await _deferredCredentialStorage.update(agent, record)
+  },
+}
+
+export const useDeferredCredentials = () => {
```

Comment:
I agree we should make this a record per item at some point. The generic records would work fine for this.Not sure why we didn't do that immediatley tbh. Probably because this was quicker 🤷